src/transformers/models/propainter/modeling_propainter.py

```diff
@@ -0,0 +1,4630 @@
+
+        _, num_channels, _, feature_height, feature_width = features_enc1.shape
+        features_enc1 = (
+            features_enc1.permute(0, 2, 1, 3, 4).contiguous().view(-1, num_channels, feature_height, feature_width)
+        )  # batch_size*timesteps num_channels height width
+        features_dec2 = self.decoder2(features_prop) + features_enc1
+
+        _, num_channels, _, feature_height, feature_width = downsample_inputs.shape
+        downsample_inputs = (
+            downsample_inputs.permute(0, 2, 1, 3, 4).contiguous().view(-1, num_channels, feature_height, feature_width)
+        )  # batch_size*timesteps num_channels height width
+
+        features_dec1 = self.decoder1(features_dec2)
+
+        flow = self.upsample(features_dec1)
+        edge = self.edgeDetector(flow)
+        edge = edge.view(batch_size, timesteps, 1, height, width)
+
+        flow = flow.view(batch_size, timesteps, 2, height, width)
+
+        return flow, edge
+
+    def forward_bidirect_flow(self, masked_flows_bi, masks):
+        """
+        Args:
+            masked_flows_bi: [masked_flows_f, masked_flows_b] | (batch_size, timesteps-1, 2, height, width), (batch_size, timesteps-1, 2, height, width)
+            masks: batch_size, timesteps, 1, height, width
+        """
+        masks_forward = masks[:, :-1, ...].contiguous()
+        masks_backward = masks[:, 1:, ...].contiguous()
+
+        # mask flow
+        masked_flows_forward = masked_flows_bi[0] * (1 - masks_forward)
+        masked_flows_backward = masked_flows_bi[1] * (1 - masks_backward)
+
+        # -- completion --
+        pred_flows_forward, pred_edges_forward = self.forward(masked_flows_forward, masks_forward)
+
+        # backward
+        masked_flows_backward = torch.flip(masked_flows_backward, dims=[1])
+        masks_backward = torch.flip(masks_backward, dims=[1])
+        pred_flows_backward, pred_edges_backward = self.forward(masked_flows_backward, masks_backward)
+        pred_flows_backward = torch.flip(pred_flows_backward, dims=[1])
+        if self.training:
+            pred_edges_backward = torch.flip(pred_edges_backward, dims=[1])
+
+        return [pred_flows_forward, pred_flows_backward], [
+            pred_edges_forward,
+            pred_edges_backward,
+        ]
+
+    def combine_flow(self, masked_flows_bi, pred_flows_bi, masks):
+        masks_forward = masks[:, :-1, ...].contiguous()
+        masks_backward = masks[:, 1:, ...].contiguous()
+
+        pred_flows_forward = pred_flows_bi[0] * masks_forward + masked_flows_bi[0] * (1 - masks_forward)
+        pred_flows_backward = pred_flows_bi[1] * masks_backward + masked_flows_bi[1] * (1 - masks_backward)
+
+        return pred_flows_forward, pred_flows_backward
+
+
+class ProPainterEncoder(nn.Module):
+    def __init__(self, config: ProPainterConfig):
+        super().__init__()
+        self.config = config
+        self.group = [1, 2, 4, 8, 1]
+        negative_slope = 0.2
```

Comment:
that's good! let's add it to the config too

Reply:
done